cdk/domino_cdk/eks_stack.py

```diff
@@ -494,6 +580,9 @@ def provision_unmanaged_nodegroup(self, name: str, cfg: dict, eks_version: eks.K
                         "--register-with-taints={}".format(",".join(["{}={}".format(k, v) for k, v in taints.items()]))
                     )
                 options["bootstrap_options"] = eks.BootstrapOptions(kubelet_extra_args=" ".join(extra_args))
+                
+                if cfg["ssm_agent"]:
+                    asg.user_data.add_on_exit_commands("yum install -y https://s3.amazonaws.com/ec2-downloads-windows/SSMAgent/latest/linux_amd64/amazon-ssm-agent.rpm")
```

Comment:
A modification of asg.user_data after it was "consumed" by LaunchTemplate at line 543 looks cheesy.

What is the effect and side effect?

Do you intend to modify asg AFTER its user data was used in LaunchTemplate?

Or do you pass an object by reference and do a backdated update of that object?

Reply:
About how it's modifying it: `asg.user_data` is an attribute of the asg that's meant to be interacted with in this fashion, and adding via the add_on_exit_commands ensure that we are adding this after the main user_data script has been ran (and that we're not interfering with it).

About why we use the attribute: We do _not_ necessarily specify `user_data`, even though there is still a `user_data` script. If `user_data` is specified manually in the config, we'll use that. Otherwise, we'll let the default `user_data` in the ami be used, in which case we don't have access to the `user_data` until after the asg is created. You'll also note that `asg.user_data` is what's input into the launch template's user_data attribute.

Reply:
Yes, I note all of this.

Let's try again. 

You do
```
if user_data:
  create_launch_template( user_data = asg.user_data )

if ssm:
  asg.user_data.append(stuff)
``` 

My grievance is: the effect of this sequence is unclear and largely depends on the internal logic of "create_launch_template".

If it copies user data, then effect of changing it in asg will affect asg only.

If it keeps user data as a reference, then effect of changing it in asg will have side effect of changing the launch template as well.

I suggest to make this side effect clear if it exists and desired.

Reply:
Also

user_data will go to both LaunchConfiguration and LaunchTemplate. 

I cannot find distinctive answer in the doc but it looks like LaunchTemplate will override LaunchConfiguration

If user_data is copied, then your SSM code will be lost. 

I'll probably check this today

Reply:
Correction :)

You do:
```
if user_data:
  create_launch_template( user_data = asg.user_data )
else:
  if ssm:
    asg.user_data.append(stuff)
```

I.e. ssm stuff is skipped if user_data is present.

I could not comprehend that this is kinda desired logic. Is it?

Reply:
Let's move on.

I will probably change this part anyway resulting from my user_data investigation

Reply:
The launch template _always_ needs to be created, as this is the unmanaged nodegroup code, and we always delete the launch config and replace it with our launch template.

So full logic is this:

```
custom_ami_id, custom_user_data = get_custom_machine_image_settings()
machine_image = get_custom_machine_image_or_return_default_eks_machine_image()
asg = create_asg(machine_image)
lt = create_launch_template()
hack_to_use_launch_template_instead_of_launch_config(asg, lt)
if not custom_user_data: # we're using standard eks user data, so safe to do normal domino customizations
  options = setup_eks_bootstrap_args()
  if cfg["ssm_agent"]:
    asg.user_data.add_on_exit_commands("...")
self.cluster.connect_auto_scaling_group_capacity(asg, options)
```

Ignoring whether you might be able to move the `if not custom_user_data` block up a few lines, _this logic should not change_, at least WRT current functionality.

Reply:
I can update the old code to try and make all of this more clear.

Reply:
ok, let's see

```
custom_ami_id, custom_user_data = get_custom_machine_image_settings()
if custom_user_data:
  machine_image = custom_machine_image( custom_user_data )
asg = create_asg(machine_image) # assuming asg.user_data is populated
lt = create_launch_template( user_data = asg.user_data )
hack_to_use_launch_template_instead_of_launch_config(asg, lt)
if not custom_user_data: # we're using standard eks user data, so safe to do normal domino customizations
  options = setup_eks_bootstrap_args()
  if cfg["ssm_agent"]:
    asg.user_data.add_on_exit_commands("...")
self.cluster.connect_auto_scaling_group_capacity(asg, options)
```
Here we make several not-so-obvious assumptions.

1. user_data from AMI will go into asg and will be immediately available as asg.user_data. I'm glad if this works. 

2. After creating the launch_template, we modify user_data in asg. But, launch_template, if specified, replaces launch_config. So my expectations here: after we have created launch_template, settings in asg.user_data are ignored. launch_template has no comment "user data is still mutable after creations" like user_data in asg.

I probably do not understand why 
1. if we use auto-assigned image, we are ok with user_data coming with the image
2. if we assign a specific image, we override user_data

Reply:
Assumption section:

1. user_data will get determined when creating the machine_image object, either it uses the user_data you provide it or it defaults to what's in the AMI. Since we pass the machine image object to the asg, that's where it gets it from.
2. I can change asg.user_data to lt.user_data if you prefer. They both do the same thing (I've tested both ways). I can also/alternatively move the entire if block (ie `if not user_data`) above the launch template definition, but I do think it's much nicer having the bootstrap info at the end. Would one of these solutions resolve this concern?

Understand section:

1. The image we're retrieving if an AMI wasn't specified is the default eks AMI, which has user_data specifically designed to bring up an EKS cluster. We need want to mess with it, it's meant to be used as-is, as if we manually specify user_data that means we now have to track the user_data of the official eks ami in case it changes.
2. If you assign a specific image, the assumption is that you're doing something like trying to get EKS to run off a stock RedHat AMI, in which case it's just flat out not going to work unless you setup user_data. We aren't trying to cover the case that someone generates an EKS-focused ami with the user_data setup. It's both extremely unlikely to ever come up, and muddies the waters for the officially supported thing vs the custom thing (and has an easy fix of "put the user_data in the config"--and the concern about user_data changing doesn't really matter here since it's specific to the specified ami).

Reply:
Probably just a comment that user data in LT is still mutable at this point.

Reply:
Thanks for the explanations about EKS vs random image

Reply:
I still do not like that ssm will not work if the user data is custom. 

I understand that we expect that whoever provides custom_user data should write the whole deal, but perhaps SSM is special.

I understand that is it tricky to add SSM to a random bash script, but EKS deals with it with Mime, merging several user_data.

I still not not exactly understand how MIME works, though

Reply:
I switched it to `lt.user_data`, and added a comment about it. I also added an error condition so if users turn on things that would have previously been ignored with custom user_data, now they'll be notified.

We really can't even assume that a machine has yum, necessarily. I think it's on the wrong side of a philosophical line to be doing any implicit/hidden user_data manipulation under the scenes in the circumstance that the user is providing it. Hopefully the error condition making things explicit is a bit better than nothing, though.